substrate/bft/src/lib.rs

```diff
@@ -304,27 +299,42 @@ impl Drop for AgreementHandle {
 
 /// The BftService kicks off the agreement process on top of any blocks it
 /// is notified of.
-pub struct BftService<P, E, I> {
+pub struct BftService<P, I> {
 	client: Arc<I>,
-	executor: E,
 	live_agreements: Mutex<HashMap<HeaderHash, AgreementHandle>>,
 	timer: Timer,
 	round_timeout_multiplier: u64,
 	key: Arc<ed25519::Pair>, // TODO: key changing over time.
 	factory: P,
 }
 
-impl<P, E, I> BftService<P, E, I>
+impl<P, I> BftService<P, I>
 	where
 		P: ProposerFactory,
-		E: Executor<BftFuture<P::Proposer, I>>,
 		I: BlockImport + Authorities,
 {
+
+	/// Create a new service instance.
+	pub fn new(client: Arc<I>, key: Arc<ed25519::Pair>, factory: P) -> BftService<P, I> {
+		BftService {
+			client: client,
+			live_agreements: Mutex::new(HashMap::new()),
+			timer: Timer::default(),
+			round_timeout_multiplier: 4,
+			key: key, // TODO: key changing over time.
+			factory: factory,
+		}
+	}
+
 	/// Signal that a valid block with the given header has been imported.
 	///
 	/// If the local signing key is an authority, this will begin the consensus process to build a
 	/// block on top of it. If the executor fails to run the future, an error will be returned.
-	pub fn build_upon(&self, header: &Header) -> Result<(), P::Error> {
+	pub fn build_upon<InStream, OutSink, E>(&self, header: &Header, input: InStream, output: OutSink, executor: E) -> Result<(), P::Error> where
```

Comment:
may as well skip the executor as a parameter if it isn't owned by the service anymore.

Reply:
Skip the executor as a parameter of what? This function?

Reply:
yeah. this function takes an executor as a parameter just so it can call `executor.execute` at the end. might as well just do that at the call sites and save some complexity